File: app/domain/market_hours.py
"""
Market hours and timezone awareness.
Called by the Risk system as the first check before any order.
NSEpy is BANNED. This module handles all exchange scheduling.
"""
from datetime import datetime, time
from typing import Literal
import pytz


SessionStatus = Literal["pre_market", "open", "after_hours", "closed"]

_SCHEDULES: dict[str, dict] = {
    "NASDAQ": {
        "tz": "America/New_York",
        "open": time(9, 30),
        "close": time(16, 0),
        "pre_market_open": time(4, 0),
        "after_hours_close": time(20, 0),
        "weekdays": {0, 1, 2, 3, 4},  # Mon–Fri
    },
    "NYSE": {
        "tz": "America/New_York",
        "open": time(9, 30),
        "close": time(16, 0),
        "pre_market_open": time(4, 0),
        "after_hours_close": time(20, 0),
        "weekdays": {0, 1, 2, 3, 4},
    },
    "NSE": {
        "tz": "Asia/Kolkata",
        "open": time(9, 15),
        "close": time(15, 30),
        "pre_market_open": time(9, 0),
        "after_hours_close": time(15, 30),  # NSE has no official after-hours
        "weekdays": {0, 1, 2, 3, 4},
    },
    "BSE": {
        "tz": "Asia/Kolkata",
        "open": time(9, 15),
        "close": time(15, 30),
        "pre_market_open": time(9, 0),
        "after_hours_close": time(15, 30),
        "weekdays": {0, 1, 2, 3, 4},
    },
    "CRYPTO": {
        "tz": "UTC",
        "open": time(0, 0),
        "close": time(23, 59),
        "pre_market_open": time(0, 0),
        "after_hours_close": time(23, 59),
        "weekdays": {0, 1, 2, 3, 4, 5, 6},  # 24/7
    },
}


def get_session_status(exchange: str, now_utc: datetime | None = None) -> SessionStatus:
    """Return current session status for the given exchange."""
    schedule = _SCHEDULES.get(exchange.upper())
    if not schedule:
        return "closed"

    tz = pytz.timezone(schedule["tz"])
    now_local = (now_utc or datetime.utcnow().replace(tzinfo=pytz.utc)).astimezone(tz)

    if now_local.weekday() not in schedule["weekdays"]:
        return "closed"

    t = now_local.time()
    if schedule["open"] <= t < schedule["close"]:
        return "open"
    if schedule["pre_market_open"] <= t < schedule["open"]:
        return "pre_market"
    if schedule["close"] <= t <= schedule["after_hours_close"]:
        return "after_hours"
    return "closed"


def is_market_open(exchange: str, now_utc: datetime | None = None) -> bool:
    """True only during regular trading hours (not pre-market or after-hours)."""
    return get_session_status(exchange, now_utc) == "open"


def exchange_timezone(exchange: str) -> str:
    return _SCHEDULES.get(exchange.upper(), {}).get("tz", "UTC")


def next_open_time(exchange: str, now_utc: datetime | None = None) -> datetime | None:
    """
    Return the UTC datetime of the next regular session open for the exchange.
    Returns None for CRYPTO (always open) and unknown exchanges.
    """
    schedule = _SCHEDULES.get(exchange.upper())
    if not schedule:
        return None
    if exchange.upper() == "CRYPTO":
        return None  # 24/7

    tz = pytz.timezone(schedule["tz"])
    base = (now_utc or datetime.utcnow().replace(tzinfo=pytz.utc)).astimezone(tz)

    open_time = schedule["open"]
    weekdays  = schedule["weekdays"]

    # Walk up to 7 days forward to find the next open
    for day_offset in range(8):
        candidate = base + __import__("datetime").timedelta(days=day_offset)
        if candidate.weekday() not in weekdays:
            continue
        local_open = candidate.replace(
            hour=open_time.hour, minute=open_time.minute, second=0, microsecond=0,
        )
        if local_open > base:
            return local_open.astimezone(pytz.utc)
    return None

File: frontend/src/lib/marketHours.ts
/**
 * Client-side market-session status, mirroring app/domain/market_hours.py.
 * Used for the "market hours status badge throughout" Phase 6 requirement.
 * This is a display convenience; the backend remains the source of truth for
 * the actual trading gate (risk.py / MarketState).
 */

export type SessionStatus = 'pre_market' | 'open' | 'after_hours' | 'closed'

interface Schedule {
  tz: string
  open: number          // minutes-from-midnight, local exchange time
  close: number
  preOpen: number
  afterClose: number
  weekdays: number[]    // JS getDay(): 0=Sun … 6=Sat
}

const hm = (h: number, m: number) => h * 60 + m
const WEEKDAYS = [1, 2, 3, 4, 5]          // Mon–Fri
const ALL_DAYS = [0, 1, 2, 3, 4, 5, 6]

const SCHEDULES: Record<string, Schedule> = {
  NASDAQ: { tz: 'America/New_York', open: hm(9, 30), close: hm(16, 0), preOpen: hm(4, 0), afterClose: hm(20, 0), weekdays: WEEKDAYS },
  NYSE:   { tz: 'America/New_York', open: hm(9, 30), close: hm(16, 0), preOpen: hm(4, 0), afterClose: hm(20, 0), weekdays: WEEKDAYS },
  NSE:    { tz: 'Asia/Kolkata',     open: hm(9, 15), close: hm(15, 30), preOpen: hm(9, 0), afterClose: hm(15, 30), weekdays: WEEKDAYS },
  BSE:    { tz: 'Asia/Kolkata',     open: hm(9, 15), close: hm(15, 30), preOpen: hm(9, 0), afterClose: hm(15, 30), weekdays: WEEKDAYS },
  CRYPTO: { tz: 'UTC',             open: hm(0, 0),  close: hm(23, 59), preOpen: hm(0, 0), afterClose: hm(23, 59), weekdays: ALL_DAYS },
}

/** Local weekday + minutes-from-midnight for a given IANA timezone. */
function localParts(tz: string, now: Date): { weekday: number; minutes: number } {
  const fmt = new Intl.DateTimeFormat('en-US', {
    timeZone: tz, weekday: 'short', hour: '2-digit', minute: '2-digit', hour12: false,
  })
  const parts = fmt.formatToParts(now)
  const get = (t: string) => parts.find((p) => p.type === t)?.value ?? ''
  const days: Record<string, number> = { Sun: 0, Mon: 1, Tue: 2, Wed: 3, Thu: 4, Fri: 5, Sat: 6 }
  let hour = parseInt(get('hour'), 10)
  if (hour === 24) hour = 0 // some engines emit "24" at midnight
  return { weekday: days[get('weekday')] ?? 0, minutes: hm(hour, parseInt(get('minute'), 10)) }
}

export function sessionStatus(exchange: string, now: Date = new Date()): SessionStatus {
  const s = SCHEDULES[exchange.toUpperCase()]
  if (!s) return 'closed'
  const { weekday, minutes } = localParts(s.tz, now)
  if (!s.weekdays.includes(weekday)) return 'closed'
  if (minutes >= s.open && minutes < s.close) return 'open'
  if (minutes >= s.preOpen && minutes < s.open) return 'pre_market'
  if (minutes >= s.close && minutes <= s.afterClose) return 'after_hours'
  return 'closed'
}

export const SESSION_LABEL: Record<SessionStatus, string> = {
  pre_market: 'Pre-Market',
  open: 'Market Open',
  after_hours: 'After Hours',
  closed: 'Market Closed',
}

export const SESSION_VARIANT: Record<SessionStatus, 'success' | 'warning' | 'muted'> = {
  open: 'success',
  pre_market: 'warning',
  after_hours: 'warning',
  closed: 'muted',
}

const DAY_NAMES = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']

/**
 * Returns a human-readable "Opens at …" string, e.g. "Today 9:30 AM ET" or "Mon 9:15 AM IST".
 * Returns null for CRYPTO (always open) or unknown exchanges.
 */
export function nextOpenLabel(exchange: string, now: Date = new Date()): string | null {
  const s = SCHEDULES[exchange.toUpperCase()]
  if (!s) return null
  if (exchange.toUpperCase() === 'CRYPTO') return null   // 24/7

  const openMinutes = s.open
  const openH = Math.floor(openMinutes / 60)
  const openM = openMinutes % 60

  // Walk up to 7 days to find next weekday the exchange is open
  for (let dayOffset = 0; dayOffset <= 7; dayOffset++) {
    const candidate = new Date(now.getTime() + dayOffset * 86_400_000)
    const { weekday } = localParts(s.tz, candidate)
    if (!s.weekdays.includes(weekday)) continue

    // Build a local-time Date at the open hour/minute for this candidate day
    // Use Intl to figure out the UTC equivalent
    const { minutes: currentMinutes } = localParts(s.tz, candidate)
    if (dayOffset === 0 && currentMinutes >= openMinutes) continue  // already past open today

    const tzAbbr = new Intl.DateTimeFormat('en-US', {
      timeZone: s.tz, timeZoneName: 'short',
    }).formatToParts(candidate).find((p) => p.type === 'timeZoneName')?.value ?? ''

    const h12 = openH % 12 || 12
    const ampm = openH < 12 ? 'AM' : 'PM'
    const mStr = openM.toString().padStart(2, '0')
    const label = dayOffset === 0 ? 'Today' : dayOffset === 1 ? 'Tomorrow' : DAY_NAMES[weekday]
    return `${label} ${h12}:${mStr} ${ampm} ${tzAbbr}`.trim()
  }
  return null
}
